### ISSUE TITLE
508-defect-1 ❗ Launchblocker [SCREENREADER]: Wizard remediation

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#22725]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** Josh

## User Story or Problem Statement

As a screen reader user, I expect to hear announcements when information is being added and removed from the page.

## Details

[As noted in issue ticket #399, dynamically revealing and hiding important information will create launchblocking usability problems](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/399) in the current state of the wizard for the following reasons:

1. There's no clear end point
2. Validation is not explicit
3. A series of dynamic, nested fieldsets are undesirable
4. We never tested this pattern on mobile or with screen readers
5. Providing results using dynamic show/hide behavior is spotty for screen reader users

### Recommended
- [ ] Transition to one thing per page

### Temporary option ACs
In order to make this accessible until the one thing per page pattern is established, we'll need to ensure:
- [ ] Screen reader users know where the form portion of the wizard ends and where the results begin
- [ ] Screen reader users know when content is added to the page

## Steps to Recreate

1. Enter `<https://staging.va.gov/manage-va-debt/request-debt-help-5655/introduction>` in browser
2. Confirm that none of the grey boxes are announced to screen reader users upon radio button selections

## Recommend Solution
Given the complexity of this particular wizard, it would be best if this were handled as one thing per page where there can be an explicit results page to the form. [The issues in ticket #399 are made worse with more content and more options](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/399).

### 1. Use `<section>` or `role="region"` to establish a clear results portion to the wizard
Programmatically placing the wizard's results within an explicit `<section>` or `role="region"` will help accomplish the first acceptance criteria of helping screen reader users know where the form portion of the wizard ends and where the results begin. 

To do so, we'll need to:
- Place the wizard results _outside_ of the `<form>`, this will semantically place it _outside_ of the form as it isn't technically a part of the `<form>`
- Nest or convert the div into a `<section>` with a `role="region"` OR attach the attribute `role="region"` to the `<div>`, this will semantically group the results information together as separate from the wizard radio options.
- Convert the `<p>` "Based on the information you provided, this isn’t the form you need." into a `<h3>` "This isn't the form you need" and give it a unique id e.g. `<h3 id="results">This isn't the form you need</h3>`
- Add `aria-labelledby` to the new region and link it to the `<h3>`, e.g. `<section role="region" aria-labelledby="results">`, this will ensure that when a user tabs out of the last radio option and into the results, the user will hear the first tabbed to link _along_ with the region's heading. For example, instead of just hearing "Find out how to pay your VA copay bill", they will hear "Find out how to pay your VA copay bill, This isn't the form you need."

### 2. Use `aria-live="polite"` to announce the dynamically added content (the grey boxes)
Screen reader users will not know when content is being added or removed from the page unless it is explicitly announced. This is important for this page in particular since results are provided in dynamic grey boxes that are added and removed from the page.

We can use `aria-live="polite"` to announce new regions that screen reader users should be aware of while making choices in the wizard. In order for this to work, the `aria-live` region should exist in the DOM on page load. It should be the parent to the region we created in step 1. e.g.:

```html
<form>
  <!-- Wizard -->
</form>
<!-- On page load -->
<div aria-live="polite" aria-atomic="true">
  <!-- Empty -->
</div>
```

Upon results being dynamically added in

```html
<form>
  <!-- Wizard -->
</form>
<!-- User selects a radio option that triggers a result -->
<div aria-live="polite" aria-atomic="true">
  <section role="region" aria-labelledby="results">
    <h3 id="results">This isn't the form you need</h3>
  </section>
</div>
```
Should the `aria-live` region not be able to exist on page load, we'll need to use a javascript delay (we can talk about that in more detail if this isn't possible.)

#### Before fixes

https://user-images.githubusercontent.com/14154792/113758317-b39ab280-96e1-11eb-8bb6-6c2bff7d337a.mov

https://user-images.githubusercontent.com/14154792/113758934-797de080-96e2-11eb-8644-763998623266.mov

#### Post fixes

https://user-images.githubusercontent.com/14154792/113754271-0e7ddb00-96dd-11eb-974b-b4990ef44b95.mov

https://user-images.githubusercontent.com/14154792/113758946-7da9fe00-96e2-11eb-99d2-c56a11198ee5.mov

## Tickets to reference

The remediation process has been handled in other accessibility tickets which may be helpful to reference. Majority of these are `defect-level-2`s as they are shorter and less complex than this particular instance.

- [Accessibility Feedback - Design intent eBenefits - CH31 Orientation Content](https://github.com/department-of-veterans-affairs/va.gov-team/issues/21478)
- [sitewide 508-defect-2 [COGNITION, SCREEN READER]: Screen reader users MUST know there is additional information after making a selection in the wizard #18581](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18581)
- [508-defect-2 [COGNITION, SCREEN READER]: Screen reader users SHOULD know there is relevant info after making a selection in the wizard](https://github.com/department-of-veterans-affairs/va.gov-team/issues/16745)
- [[SCREENREADER]: Text and inputs added dynamically should be announced by screen readers](https://github.com/department-of-veterans-affairs/va.gov-team/issues/316)

## Additional notes
This should be fixed before launch as this presents a significant usability challenge to screen reader users. Should that not be possible due to time or business constraints, we can work to find alternative temporary solutions (e.g. providing instructions or explanation of the interactions to the user before entering the wizard so they know to expect results post selections.)

**Comment by LinnJS** [comment]
@joshkimux this work is currently in prod and staging if you want to test it manually before closing this ticket

**Comment by joshkimux** [comment]
@LinnJS stellar work!! 🥳 🎈 
I've tested this and results are being announced as expected. Upon jumping to elements in the grey results box, screen reader users are also notified that they are in a new region!

https://user-images.githubusercontent.com/14154792/115571445-ff765b80-a28c-11eb-9676-ed3c00a352cc.mov

That being said, two final (hopefully minor) changes:

## For some results, the grey box does not appear
<img width="911" alt="Screen Shot 2021-04-21 at 10 31 57 AM" src="https://user-images.githubusercontent.com/14154792/115571649-32205400-a28d-11eb-948b-b85bddd9111e.png">

## If possible, I'd recommend making the new `h2` look like a `h2`
You might need blessings from your designer, but this will make it more obvious to users with low vision and cognitive disabilities that they are viewing results. All you'd need to do is remove the `vads-ufont-size--h6`, `vads-u-font-weight--normal`, and  `vads-u-font-family--sans` classes 😄 

For example:
```html
<h2 class="vads-u-margin-top--0" id="wizard-results">Based on the information you provided, this isn’t the form you need.</h2>
```

If it's too big for y'alls taste, adjusting to a h3 sizing could work as well
```html
<h2 class="vads-u-margin-top--0 vads-u-font-size--h3" id="wizard-results">Based on the information you provided, this isn’t the form you need.</h2>
```

<img width="834" alt="Screen Shot 2021-04-21 at 10 39 13 AM" src="https://user-images.githubusercontent.com/14154792/115572358-d1dde200-a28d-11eb-8e97-7a5f02032616.png">

**Comment by denisecoveyduc** [comment]
As per comments above this has been resolved!
